### ISSUE TITLE
A record with an enum as key resolves to TObject<{}>

### ISSUE THREAD

**Issue by dippi** [sinclairzx81/typebox#564]
# Issue

```typescript
enum E {
  A = 'A',
  B = 'B',
  C = 'C',
};
const T = Type.Record(Type.Enum(E), Type.Number());

```
`T` has type `TObject<{}>` instead of `TObject<{ A: TNumber; B: TNumber; C: TNumber }>`

This is because `Type.Enum(E)` is equivalent to `TUnion<E[]>` instead of `TUnion<['A', 'B', 'C']>` and in `TRecordFromUnionRest` it doesn't match `K extends [infer L, ...infer R]`.

https://github.com/sinclairzx81/typebox/blob/010ff1840c7ab43d3df6a52f922e365184635217/src/typebox.ts#L653-L657

# Possible fix

I tried adding `K extends TEnum<infer S> ? TObject<{ [K in keyof S]: T }> :` as first case of `TRecordResolve` and it seems to be working as expected.

https://github.com/sinclairzx81/typebox/blob/010ff1840c7ab43d3df6a52f922e365184635217/src/typebox.ts#L671-L679

# Regression tests

I'd be happy to make a PR, but I had an issue writing a regression test.

I tried to add the example above in [/test/static/record.ts](https://github.com/sinclairzx81/typebox/blob/18d1cf7ac57c8b5f0292cb02ddbd1707e86c8c4c/test/static/record.ts) with the assertion `Expect(T).ToStatic<Record<E, string>>();`, but it succeeds even though it shouldn't.

The reason is that `.ToStatic<U>()` only checks that `U` extends `Static<T>` instead of checking equality (and `{}` is super-type of `Record` so it always succeeds).

In another project I took inspiration from [gvergnaud/ts-pattern](https://github.com/gvergnaud/ts-pattern/blob/435284ac08cd0c4c1361fd59ae936bd1ece7c85b/src/types/helpers.ts#L67-L71) to assert type equality and I could use it for this specific case.

However I think the best case scenario would be to somehow evolve `.ToStatic` so that the whole test suite would benefit from it, but I don't know how.

Let me know if you have any suggestion.

# Additional context

We updated from `0.29.4` to `0.31.5`, but I also checked later versions and the issue persists.

**Comment by sinclairzx81** [comment]
@dippi Hi, Thanks for reporting! (and for the detailed write up)

Let me have a look into this. I think a solution may be to drop `TEnum` as a `interface` type, and reimplement as an alias type that maps directly into `TUnion<[...]>` (as this is the actual representation for enum). Quite a few other types do this already (such as KeyOf), but enum is a bit of special case.

By remapping Enum into Union, this would resolve Record Key mappings with no additional updates required to Record inference (which is very complex)

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAsAChZJCOxbwA0nAC8KYQDoAqu2CKAFAG1ZcLZm0AZYDAxQCXIwwCCDAJRZzljDbsOTi4AQp7eMhZoVrb2js4MAMJhsgC6HrLyispwAEoavto5GApQACZGKjhRfgByAK4gIg5GHuly7UKYufn4RKQAPJ0YEFS5AHxw-PxwQ92aCD4WS8srq8tTS64AXHDsDU1QANyLa6enGxbBO3uNDscRZ4+rF3AJ1-t3J09PG+SyQA)

```typescript
import { Type, Static } from '@sinclair/typebox'

const K = Type.Union([
  Type.Literal('A'),
  Type.Literal('B'),
  Type.Literal('C'),
])

const R = Type.Record(K, Type.Number())


type R = Static<typeof R> // type R = {
                          //   A: number;
                          //   B: number;
                          //   C: number;
                          // }
```
Let me investigate
Cheers
S

**Comment by dippi** [comment]
Hi @sinclairzx81, thanks for the quick reply!

I agree, it would be great to solve the issue upstream at `TEnum` definition. However I must say I liked that `TEnum` didn't depend on `UnionToTuple` like `KeyOf` does, because it's a [risky conversion](https://stackoverflow.com/a/55128956).

Is there any chance that we could adapt `TUnion<('A' | 'B' | 'C')[]>` to work like `TUnion<['A', 'B', 'C']>`? I think that would fix it as well, because `TEnum<E> extends TUnion<infer U>` is already satisfied and `U` is inferred as `E[]`, which should be equivalent to the array of the union of the enum values.

I'm not sure if I was able to explain the idea, let me know if there's anything I can clarify!

**Comment by sinclairzx81** [comment]
@dippi Heya,

Unfortunately, I'm not sure I'm able to implement much here as TypeScript still treats enum values as opaque (so I'm unable to map them into unions). Just checked the current state of things, and the situation is still largely the same. 

The following is the main issue.

```typescript
enum E {
  A = 1,
  B = 2
}

type X = keyof typeof E // type X = 'A' | 'B'
type Y = typeof E[X]    // type Y = E - Expect 1 | 2
```
Note that the values `1 | 2` would need to be mapped for the union, not `'A' | 'B'`. So if it is somehow possible to extract the enum values (as a union) then I can continue to implement a union mapping (resolving this issue)

### UnionToTuple

Agree about the usage of this mapping. However it's use is generally ok if applied to constant literal values in contexts where you expect a distinct set (such as mapping property keys on objects where each key is unique). Although, I am actually trying to limit the usage of this type (restricted to only key distributions), but it's a bit tricky on some of the more sophisticated mappings (TComposite especially)

```typescript
type T = 'a' | 'a' | 'a' | 'b' // type T = "a" | "b"

type U = UnionToTuple<T>       // type U = ['a', 'b']
```

I'll leave this issue open for a day and do a bit more digging. In instances where users do not define explicit values for the enum variants, it would be possible to map ordinal values via [this](https://github.com/sinclairzx81/typebox/blob/master/examples/typedef/typedef.ts#L36-L42) approach, but without being able to map the user defined values, I'd need to keep the functionality "as is" (which means using enums for keys on records would have to remain a known bug / limitation)

Let me know if you know of any tricks to map out these values.
Cheers
S

**Comment by dippi** [comment]
> I'm not sure I'm able to implement much here as TypeScript still treats enum values as opaque (so I'm unable to map them into unions)

I see that `Y = E` instead of `1 | 2`, but it seems to me that they are behaving mostly the same. Here some example with `Exclude` or mapping each type of enum / union.

```typescript
import { TLiteral, TLiteralValue } from '@sinclair/typebox'

enum E {
  A = 1,
  B = 2,
  C = 3
}

type WrapIt<T extends TLiteralValue> = T extends any ? TLiteral<T> : never;

type X = keyof typeof E // type X = 'A' | 'B' | 'C'
type Y = typeof E[X]    // type Y = E
type W = Exclude<E, 1>  // type Y = E.B | E.C
type Z = WrapIt<E>      // type Z = TLiteral<E.A> | TLiteral<E.B> | TLiteral<E.C>
```

> I am actually trying to limit the usage of this type (restricted to only key distributions), but it's a bit tricky on some of the more sophisticated mappings (TComposite especially)

I think that using `UnionToTuple` in `TComposite` is fine, as it's an hidden detail. In case of `TKeysOf` instead the resulting tuple is "leaked" into `TUnion<...>` and, even if unlikely, somebody could access `anyOf` and find the array elements in a different arrangement.

> I'd need to keep the functionality "as is" (which means using enums for keys on records would have to remain a known bug / limitation)

If the other attempts don't pan out, would you consider to accept the patch to `TRecordResolve` that I originally suggested?

**Comment by sinclairzx81** [comment]
@dippi Heya,

> If the other attempts don't pan out, would you consider to accept the patch to TRecordResolve that I originally suggested?

Would be keen to pursue the Enum to Union mapping rather than mappings on TRecordResolve, This mostly to keep the complexity down on that type. 

> I see that Y = E instead of 1 | 2, but it seems to me that they are behaving mostly the same. Here some example with Exclude or mapping each type of enum / union.

You know, I think you're onto on to something special here. TS is able to extract out the enum values when assigned as property keys.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKZwL5wGZQQhwDkAAgM7AB2AxgDYCGwUA9DOhgEYQAeJAWABQwjNQCuxAKKI4AQTgBeUgAsM9ehBIAaOACElcACwAmXQGE4OYcNoRqFeMkNpMAOgDyXAFYZaMAAokKxDQsKtWVjg7BycALhQvX38AHgRhKwBtKTc5AF0E1ww3AGUYKBoAcwCASl1whpDIqzUNCEKyiupKgG4MuGy3PQKUTlLyqtr6xobm4xMOie6+oSyc8xGi8a7qupnGuaMAZkWdlewa-avrsObsAD4gA)

```typescript
import { Type } from '@sinclair/typebox'

enum E { A = 'hello', B = 42, C  }

const T = Type.Object({                 // const T: TObject<{
  [E.A]: Type.String(),                 //  hello: TString;
  [E.B]: Type.String(),                 //  42: TString;
  [E.C]: Type.String(),                 //  43: TString;
})                                      // }>
```
Based on this, here's a candidate update for TEnum implemented as an alias mapping for TUnion. Note the implementation needs to solve for the following 3 cases (Enum case is C). It seems good.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxjriQBYYgEDyYxEAdgM47IAywMGUBAGw7de-AQDVBAV2xwAqs2AtkEZFLADZyBUuY4Agq1Z8YAJQysYOAKJspUDHAC+cAGZQIIOAHIAAq2BmEgECYCgAehh0DAAjCAAPbwBYAChU8PC4AFoc3Lz8gsKi4pLS9MzkWykvAAoAOQx4+HMAN2AAlgBKcuzSvv6BwdTGyFg4KMwUKpAAaQxUOABeOEsoQIBzYfjR+AnHSuZqyQEZJZWYNeZ1uAAfOEOQGL4tnfHoqYeVET5BHRYAHmQcEavGYABNWChvmJjjIAHxnIEgjDgyEEZgLAD8UJ4PwEgIRAC57hgWs8UiNoLt3gdql9cWI1BoMIDgU0URCcaJBLCMAjln9mComZpAdN6dyBIKCXCXlS3pNaSBWcjUVy8bz+XBbKx7CztIoAYZjLBzJYxZ8INDBCL9XD7bK0ikwRhggQHG4pEEmMxtQ9-uI2aDOUreRwgxzIeYSNAwRbqnNUDhxPaariQMTkDgIIxdKxMcSKNRaAwfaxOpnpoCANoAa3mEFcKAAunCANypHqDbs93s9ZAWGCsLu90djoqdlJIDJwADCBGMcH0xMFcGtAlScDgMbY8FniOiADpBTVq5utygj+uat5qAIBBBvJ0sOet2hMIfrwAWABMz9fl4ft+ADs3QpFuzZgVuO6WHApgHh+0axjUs4cEe+CXOsNSdFBCqOPByz4EQpD-HsjZwQiM4QLWqROJO06ZPOi4AELEnQMQAFauvA65wAACkQojMOeMHwNYZzTDUCAAfExK3hg96Pi+4EXqgxK-spF5wAAXupACs55OLholwQhGCHkhUBgjU1hoR+GEbNhuF7KZhGEMQJCkdE5GmJRmTUbR9FwDOTGOLOxLTPxgl8MJKkotUcAAGIQBAiAAfoZzyYp3iaRezFnBpAGzoZIksLB4nLJJyUQMZZXNGZFmushtmAeZDlXE554uQReDuSRZFNr5wX+TRKROEAA)
```typescript
import { Type, Static, SchemaOptions, TLiteral, TLiteralValue, UnionToTuple, TUnion, AssertRest, Ensure } from '@sinclair/typebox'

// ------------------------------------------------------------------
// TEnum (Next Revision)
// ------------------------------------------------------------------
export type TEnumKey = string
export type TEnumValue = string | number
export type TEnumToLiteralUnion<T extends TLiteralValue> = T extends any ? TLiteral<T> : never
export type TEnumToLiteralTuple<T extends TLiteralValue> = UnionToTuple<TEnumToLiteralUnion<T>>
export type TEnum<T extends TLiteralValue> = Ensure<TUnion<AssertRest<TEnumToLiteralTuple<T>>>>

declare function Enum<V extends TEnumValue, T extends Record<TEnumKey, V>>(item: T, options?: SchemaOptions): TEnum<T[keyof T]>;

// ------------------------------------------------------------------
// Tests
// ------------------------------------------------------------------

{ // Case A: Union Literal
  const C = Type.Union([
    Type.Literal('hello'),
    Type.Literal(42),
    Type.Literal(47)
  ])
  const R = Type.Record(C, Type.String())
  type R = Static<typeof R> // ok
}

{ // Case B: Object Literal Pattern
  const E = Enum({
    x: 'hello',
    y: 42,
    z: 45
  })
  const R = Type.Record(E, Type.String())
  type R = Static<typeof R> // ok
}

{ // Case C: Enum Pattern
  enum Foo {
    A = 'hello',
    B = 42,
    C
  }
  const E = Enum(Foo)
  const R = Type.Record(E, Type.String())
  type R = Static<typeof R> // ok
}
```
Open to thoughts on the implementation. I'm trying to imagine a scenario where the above could break (It needs testing). Will setup a branch for review.

**Comment by sinclairzx81** [comment]
@dippi Refer to https://github.com/sinclairzx81/typebox/commit/427e735b330af292d777393f51f21277fc423956 (branch `enum`)

**Comment by dippi** [comment]
I checked the code and it looks good!

> I'm trying to imagine a scenario where the above could break

I can't think of anything :thinking: 
I also checked `Static<typeof E>` in case C and it's `Foo` as expected.

I think we can move forward with this implementation, is there anything that I can do to help? (e.g. brainstorm the adoption of type equality assertion in static tests)

**Comment by sinclairzx81** [comment]
@dippi Hi, thanks for taking a look. 

Yeah, I'm actually quite happy with the implementation and will likely move forward with it. But I do want to carry out some more testing over the weekend (as generally pushing inference changes can have unforeseen breakage), but I think it's ok and will likely publish in the next revision.

> I think we can move forward with this implementation, is there anything that I can do to help? (e.g. brainstorm the adoption of type equality assertion in static tests)

Yes actually. Currently the `ToStatic` tests only assert for the assignability, but would be open to additional tests to assert exact equality. Let me create a PR for the current work now, and if you would like to target to the `enum` branch, can look at some additional assertions when writing some tests for the new Enum type.

Also, I'm actually very keen to have you show as contributor (as I don't think this feature would be considered without your inputs and suggestions regarding enum variant union types). If you can push something to the `enum` branch, or possibly submit the additional testing infrastructure via other PR, I'd be glad to accept to get you showing up as contributor. 

Will setup a PR now.

**Comment by sinclairzx81** [comment]
@dippi See https://github.com/sinclairzx81/typebox/pull/567 for PR. If you want to drop in some quick tests around this (via PR), that would be amazing :)

**Comment by dippi** [comment]
@sinclairzx81 I made some experiments making `.ToStatic` more strict with some promising results, I'll try to push something tomorrow.

**Comment by sinclairzx81** [comment]
@dippi Hiya!

Updates published on 0.31.9. Hey, thanks again for your help with this one, the new static type asserts are amazing and the new enum functionality wouldn't have been possible without your inputs and insights. Your contributions here are very much appreciated :)

Will close up this issue for now, but feel free to ping an issue if you notice anything a miss with the new changes.
All the best!
S

**Comment by dippi** [comment]
Thank you, @sinclairzx81, for maintaining such an awesome library!
Also for being so responsive and friendly :)
